src/Data/Singletons/Single/Monad.hs

```diff
@@ -154,3 +171,66 @@ singDecsM :: DsMonad q => [Dec] -> SgM [DDec] -> q [DDec]
 singDecsM locals thing = do
   (decs1, decs2) <- singM locals thing
   return $ decs1 ++ decs2
+
+{-
+Note [Explicitly binding kind variables]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+singletons is careful to generate code that explicitly quantifies every kind
+variable bound by a top-level forall. For example, if we were to single the
+identity function:
+
+  identity :: forall a. a -> a
+  identity x = x
+
+We would first promote the types to kinds and generate an explicit forall
+mapping fresh type variables to the promoted types (step 1):
+
+  sIdentity :: forall (x :: a). Sing x -> Sing (Identity x)
+  sIdentity sX = sX
+
+Now, because we want to explicitly bind the kind variable `a`, we traverse
+the promoted types, extract the set of all kind variables, and put
+them at the front of the forall, like so (step 2):
```

Comment:
In this case, we don't need a 2-step process. You have the `forall a` in the original code. So we know we're going to need to quantify the kind `a` in the singled code. However, in the event that the original code *doesn't* have an explicit quantification, we'd need something like this. So, either you track in-scope variables in the original source so that you can recreate GHC's implicit quantification, or you do it the way it's done here. To be honest, I slightly prefer the first way -- it's conceptually closer to the way implicit quantification works -- but this approach is fine too.

Reply:
Hm, I think the Note may have been misleading, since you're describing I problem which I believe doesn't exist. You're worrying about what happens when `identity` is written as `identity :: a -> a`, but there's no difference in behavior from `singletons` end when `identity` is written with an explicit `forall` or not—either way, it'll be singled to `sIdentity` with an explicit `forall`. (Perhaps I'm misreading this comment, though...)

Reply:
I think you're misreading the comment.

Suppose all quantification were explicit. Then, we could simply echo the quantification that the user wrote -- end of story. No need to look for/track free variables.

Of course, all quantification isn't explicit. But we can do a pre-pass over the AST and make it explicit, according to simple rules. Because we already have to take a pass over the AST during promotion, this explicit quantification could happen there.

Does that clarify?

Reply:
> Of course, all quantification isn't explicit.

You keep saying this, but this simply isn't true in the context of `singletons` (and I'm talking about the current release, not just after this patch). When we single a definition,we _always_ explicitly quantify the variables, regardless of whether the original definition was explicitly quantified or not. As evidence, you can see for yourself that this definition:

```haskell
$(singletons [d| identity :: a -> a; identity x = x|])
```

Singles to:

```haskell
    sIdentity ::
      forall (t_a6BO :: a_a6BG).
      Sing t_a6BO -> Sing (Apply IdentitySym0 t_a6BO :: a_a6BG)
    sIdentity (sX :: Sing x_a6BN) = sX
```

In fact, you get the exact same singled definition when you explicitly quantify the `a` in the original source:

```haskell
$(singletons [d| identity :: forall a. a -> a; identity x = x|])
```

So I don't understand what the fuss is about.

Reply:
I'm talking about explicit quantification in the original, user-written source, not in our output. If the original, user-written source had full explicit quantification, all those type variables become the kind variables we need to quantify over in our output. So we could the quantification in the original, user-written source to guide the quantification in the output.

Reply:
> If the original, user-written source had full explicit quantification, all those type variables become the kind variables we need to quantify over in our output. So we could the quantification in the original, user-written source to guide the quantification in the output.

I mean, sure, that's one thing you could do. But I don't see that as making the implementation simpler. As you noted in https://github.com/goldfirere/singletons/pull/298#discussion_r165858547, the original source isn't always going to have explicit quantification, so we'd still need to do all of this free-variable calculation, but elsewhere in the code.

What's worse, storing the bound variables in `ADxxx` isn't always going to work, because not all of the types that we single are going to come from promotion. See this, for example:

https://github.com/goldfirere/singletons/blob/272f678dd83271a18d7ed3c181fe0f7b6395f3e8/src/Data/Singletons/Single.hs#L373-L380

where we have a scenario where we call `singType` on a type taken directly from `dsReify` (instead of through the promotion pipeline). If anything, we'd need even _more_ bookkeeping in order to propagate the right information through to code like this.

Reply:
OK, if you say so. No further objections from me.